### PR TITLE
Address suggestions from the meeting of Jan 13

### PR DESCRIPTION
**Pull request by bobsilverberg**
Make base_url optional, and reorder arguments to place selenium first
Rename _url to URL_TEMPLATE
Rename url to target_url
Prepend base_url to the result of URL_TEMPLATE if it does not result in an absolute URL
Rename kwargs to url_kwargs
Remove timeout property and replace with an argument to WebView.__init__
Move find_element* methods into Page and Region
In wait_for_page_to_load, only perform the URL check if target_url yields a value
Add/change tests for new/changed functionality

### CODE REVIEW

**Comment by bobsilverberg** [comment]
@davehunt These are ready for your review :)

I notice a coverage issue, which I will look at.

**Comment by davehunt** [comment]
This is a little hard to review because there's so much changing. Could we at least get separate commits, if not separate pull requests? It will be easier then to discuss and iterate on them as I'm sure there's more bikeshedding to do. :smile: :bike: :house: :zap:

File: pypom/view.py [hunk 1]
```diff
@@ -12,66 +12,30 @@ class WebView(object):
...
-        :param selenium:
-            An instance of the Selenium class.
+            Defaults to ``None``.
+
+        :param timeout:
+            The timeout, in seconds, to be used for calls to ``self.wait``.
+
+            Defaults to ``0``.
 
         :param kwargs:
```
Comment (davehunt) [inline pypom/view.py]: This documentation needs updating.

File: pypom/region.py [hunk 2]
```diff
@@ -29,7 +29,10 @@ def __init__(self, page, root=None, **kwargs):
         :param kwargs:
             Dictionary of arguments to pass into the parent's ``__init__``.
```
Comment (davehunt) [inline pypom/region.py]: I don't think we need kwargs here at all.

File: pypom/page.py [hunk 3]
```diff
@@ -2,19 +2,63 @@
...
 
-    _url = '{base_url}'
+    URL_TEMPLATE = ''
     """String representing a URL that will return this page.
 
     This string is formatted and can contain names of keyword arguments passed
-    during construcion of the page object. The ``{base_url}`` keyword argument
-    is always passed to the format operation.
+    during construction of the page object. The template should either assume
+    that the value of ``base_url`` will be prepended to the result of the
```
Comment (davehunt) [inline pypom/page.py]: Nit: I know I used the term 'prepended' but I think it makes it hard to parse. If we switch the order and use 'append' or 'concatenate' I think it improves the docs.

File: pypom/page.py [hunk 4]
```diff
@@ -2,19 +2,63 @@
...
+    format, or should yield an absolute URL.
     """
 
+    @property
+    def target_url(self):
```
Comment (davehunt) [inline pypom/page.py]: I still feel like 'target' is misleading. It's more of a source than a target, at least when used by `open`. How do you feel about `canonical_url`?

File: pypom/page.py [hunk 5]
```diff
@@ -2,19 +2,63 @@
...
+        Returns the URL to the current page,
+        formatted from :py:data:`URL_TEMPLATE`.
+        Unless URL_TEMPLATE results in an absolute URL, ``self.base_url`` is
+        prepended to the resulting URL.
+
+        :returns:
+            String representing a URL that will return this page.
+        """
+        return urljoin(self.base_url,
+                       self.URL_TEMPLATE.format(**self.url_kwargs))
```
Comment (davehunt) [inline pypom/page.py]: So does `urljoin` return a relative URL when `base_url` is `None`?

File: pypom/page.py [hunk 6]
```diff
@@ -23,24 +67,15 @@ def open(self):
...
-            String representing a URL that will return this page.
-        """
-        return self._url.format(base_url=self.base_url, **self.kwargs)
-
     def wait_for_page_to_load(self):
         """
         Waits for the page to load by waiting until the URL reported by
-        Selenium is the same as that returned by the :py:func:`url` property.
+        Selenium is the same as that returned by the :py:func:`url` property,
+        if the :py:func:`url` property has a value.
```
Comment (davehunt) [inline pypom/page.py]: We no longer have a `url` property.

File: tests/test_page.py [hunk 7]
```diff
@@ -46,6 +68,12 @@ def test_wait_for_page_timeout(page, selenium):
         page.wait_for_page_to_load()
 
 
+def test_wait_for_page_empty_base_url(page, selenium):
+    selenium.current_url = str(random.random())
+    page.base_url = None
```
Comment (davehunt) [inline tests/test_page.py]: Rather than setting the `base_url` to `None` could we instantiate a new `Page` without the argument? That would more closely reflect actual use of the package.

**Comment by davehunt** [comment]
You can probably ignore my first comment, I was able to roughly distinguish the changes and have given a review.

**Comment by bobsilverberg** [comment]
Thanks for the review @davehunt. As the changes were mostly interrelated I didn't want to try to make atomic commits for each. I also thought that the changes were few and simple enough to justify a single commit. I'm sorry if that was not the case and I'm glad you were able to review the PR regardless.

I will address your comments later this morning.

**Comment by bobsilverberg** [comment]
@davehunt The coverage issue seems to be with the new import path for Python 3. As the coverage is currently only generated for Python 2 it doesn't see that path as covered. I am currently experimenting with trying to get coverage reported for combined runs of Python 2 and 3.

**Comment by bobsilverberg** [comment]
@davehunt My various attempts to do this can be seen at https://github.com/bobsilverberg/PyPOM/pull/1, none of which seem to work. Do you have any thoughts about how we can get coverage reported for two different versions of Python, combined?

**Comment by bobsilverberg** [comment]
@davehunt It looks like maybe we need to do something similar to what is done in https://github.com/ionelmc/python-nameless, although that looks pretty complicated. What do you think of that?

I suppose another option, if we don't think it's that important, is to just ignore this reported coverage deficit, although fixing it and getting accurate coverage across Python versions seems like the correct approach.

**Comment by davehunt** [comment]
This is all because we're using urljoin? Can't we just concatenate the two strings instead?

File: pypom/region.py (hunk 2 above)
Comment (bobsilverberg) [inline pypom/region.py]: Good call.

File: pypom/page.py (hunk 3 above)
Comment (bobsilverberg) [inline pypom/page.py]: Okay, I've tried to improve it. We'll see what you think.

File: pypom/page.py (hunk 4 above)
Comment (bobsilverberg) [inline pypom/page.py]: I agree that `canonical_url` is more accurate, but also more confusing to people who don't know what canonical means. As we are going to be the primary consumers I am fine with it.

File: pypom/page.py (hunk 5 above)
Comment (bobsilverberg) [inline pypom/page.py]: What I like about `urljoin` is that it has URL-smarts built in. It treats the first URL as the default for all unspecified parts of the second URL, so if the second URL is already absolute it just ignores the first URL. If the first URL is empty then it will ignore that and return the second URL unchanged. I believe it does exactly what we want, but it is causing us this coverage issue because of the two imports.

I'm going to change the code to just check for the presence of `//` in the result of the format and if that is present then assume we already have an absolute URL.

File: pypom/page.py (hunk 6 above)
Comment (bobsilverberg) [inline pypom/page.py]: Good catch, thanks!

File: tests/test_page.py (hunk 7 above)
Comment (bobsilverberg) [inline tests/test_page.py]: Done.

**Comment by bobsilverberg** [comment]
@davehunt This is ready for another review.

File: pypom/page.py [hunk 8]
```diff
@@ -2,45 +2,81 @@
...
 
         :returns:
-            The current page object (i.e., ``self``).
+            String representing a URL that will return this page.
         """
-        self.selenium.get(self.url)
-        self.wait_for_page_to_load()
-        return self
+        url = self.URL_TEMPLATE.format(**self.url_kwargs)
+        return ('//' in url and url) or (self.base_url and self.base_url + url)
```
Comment (davehunt) [inline pypom/page.py]: Not sure I understand this. Doesn't this mean a URL like `/path//path` wouldn't be appended to the base URL? I know it's unlikely but it's a valid URL path.

Comment (bobsilverberg) [inline pypom/page.py]: Correct. That wouldn't get appended. It doesn't seem like a valid URL path to me, but I guess it is. I suppose if we really want to get this perfect we need a regex that allows for `http://`, `https://` and just `//`. Can you suggest anything else?

**Comment by davehunt** [comment]
I have an outstanding question on the canonical URL construction, other than that it looks fine. Going back to the original coverage concern, I am fine with us having a drop in coverage if it's just coveralls interpretation of the coverage due to the Python version. We know the tests are run in the supported Python versions, so it's not really an issue. Maybe it's worth it if `urljoin` really solves some problems for us.

**Comment by bobsilverberg** [comment]
@davehunt I responded to your question above. I'll let you decide. I'm fine with using a regex if that's what we need. I do think `urljoin` is a pretty neat solution to the problem though, and I'm also okay with having the reported coverage loss.  What do you think?

**Comment by davehunt** [comment]
Let's just 'lose' the coverage. We understand it, and know that there really is no loss in coverage.

**Comment by bobsilverberg** [comment]
Okay, @davehunt, I switched back to `urljoin`. Would you care to take a final look? I will squash the commits and merge on your r+

File: tests/test_page.py [hunk 9]
```diff
@@ -10,23 +10,48 @@
...
+
+    class MyPage(Page):
+        URL_TEMPLATE = url_template
+    page = MyPage(selenium, base_url)
+    assert url_template == page.canonical_url
+
+
+def test_canonical_url_empty(selenium):
+    page = Page(selenium)
+    assert page.canonical_url == ''
```
Comment (davehunt) [inline tests/test_page.py]: I feel like this should be None rather than an empty string, but it's just a feeling. :smile:

File: tests/test_page.py [hunk 10]
```diff
@@ -10,23 +10,48 @@
...
-        _url = '{key}'
-    page = MyPage('foo', selenium, key=value)
-    assert value == page.url
+        URL_TEMPLATE = '{key}'
+    page = MyPage(selenium, base_url, key=value)
+    assert base_url + value == page.canonical_url
+
+
+def test_canonical_url_prepend(base_url, page, selenium):
+    assert base_url == page.canonical_url
```
Comment (davehunt) [inline tests/test_page.py]: I don't think this is testing that the canonical URL is being appended to the base URL. It looks like the URL_TEMPLATE is empty and so we're just returning the base URL. To test concatenation we'd need a path for URL_TEMPLATE.

File: tests/test_page.py [hunk 11]
```diff
@@ -10,23 +10,48 @@
...
+    page = MyPage(selenium, base_url)
+    assert url_template == page.canonical_url
+
+
+def test_canonical_url_empty(selenium):
+    page = Page(selenium)
+    assert page.canonical_url == ''
+
+
+def test_canonical_url_keywords(base_url, selenium):
```
Comment (davehunt) [inline tests/test_page.py]: Perhaps excessive, but we could test the keyword arguments are expanded when the URL_TEMPLATE is an absolute URL.

**Comment by davehunt** [comment]
I think this is good to merge, but I'll give you a chance to read my few comments and see if you want to make any further changes.

File: tests/test_page.py (hunk 9 above)
Comment (bobsilverberg) [inline tests/test_page.py]: Its an empty string because of my use of `urljoin`. I simply return the result of that, which is an empty string if both `self.base_url` and `self.URL_TEMPLATE` are empty. I could change that code to return `None` if the result of `urljoin` is empty. Would you prefer that?

File: tests/test_page.py (hunk 10 above)
Comment (bobsilverberg) [inline tests/test_page.py]: I disagree that it isn't testing prepending. If prepending wasn't working then we wouldn't get the `base_url` back from `canonical_url`. Although I agree that it theoretically could be appending, or replacing, so something that uses a `URL_TEMPLATE` would be better. Having said that, `test_canonical_url_keywords` already does that, I believe. It uses a `URL_TEMPLATE` and a `base_url`, so I think we already have the test you mention. Perhaps I should just remove this test and we can rely on `test_canonical_url_keywords` to test both base_url prepending and keyword replacement? What do you think?

File: tests/test_page.py (hunk 11 above)
Comment (bobsilverberg) [inline tests/test_page.py]: It does seem excessive, especially knowing the implementation, but we shouldn't write tests based on our knowledge of the implementation, so I agree that the extra coverage might be worthwhile in case the implementation changes in the future and a possible bug is introduced. I'll add a test for it.

File: tests/test_page.py (hunk 10 above)
Comment (davehunt) [inline tests/test_page.py]: Well, it's appending the base URL to '' isn't it? Although I agree this is tested in `test_canonical_url_keywords` I think it's worth having a test with a base URL and a simple path.

File: tests/test_page.py (hunk 9 above)
Comment (davehunt) [inline tests/test_page.py]: Am I being overly picky? Perhaps `URL_TEMPLATE` could default to `None` and then in `canonical_url` we could do:

```python
if self.URL_TEMPLATE is not None:
    return urljoin(self.base_url, self.URL_TEMPLATE.format(**self.url_kwargs))
return base_url
```

**Comment by bobsilverberg** [comment]
Okay, @davehunt, it might finally be ready! Please take another look.

**Comment by davehunt** [comment]
r+ please squash and I'll do the merge

**Comment by bobsilverberg** [comment]
Squashed and pushed!

**Comment by davehunt** [comment]
...and merged! Thanks!
